sled-agent/src/long_running_tasks.rs

```diff
@@ -0,0 +1,234 @@
+//! by both the bootstrap agent and the sled-agent.
+//!
+//! We don't bother keeping track of the spawned tasks handles because we know
+//! these tasks are supposed to run forever, and they can shutdown if their
+//! handles are dropped.
+
+use crate::bootstrap::bootstore_setup::{
+    new_bootstore_config, poll_ddmd_for_bootstore_peer_update,
+};
+use crate::bootstrap::secret_retriever::LrtqOrHardcodedSecretRetriever;
+use crate::config::Config;
+use crate::hardware_monitor::{HardwareMonitor, HardwareMonitorHandle};
+use crate::storage_monitor::{StorageMonitor, StorageMonitorHandle};
+use crate::zone_bundle::{CleanupContext, ZoneBundler};
+use bootstore::schemes::v0 as bootstore;
+use key_manager::{KeyManager, StorageKeyRequester};
+use sled_hardware::{HardwareManager, SledMode};
+use sled_storage::disk::SyntheticDisk;
+use sled_storage::manager::{StorageHandle, StorageManager};
+use slog::{info, Logger};
+use std::net::Ipv6Addr;
+
+/// A mechanism for interacting with all long running tasks that can be shared
+/// between the bootstrap-agent and sled-agent code.
+#[derive(Clone)]
+pub struct LongRunningTaskHandles {
+    /// A mechanism for retrieving storage keys. This interacts with the
+    /// [`KeyManager`] task. In the future, there may be other handles for
+    /// retrieving different types of keys. Separating the handles limits the
+    /// access for a given key type to the code that holds the handle.
+    pub storage_key_requester: StorageKeyRequester,
+
+    /// A mechanism for talking to the [`StorageManager`] which is responsible
+    /// for establishing zpools on disks and managing their datasets.
+    pub storage_manager: StorageHandle,
+
+    /// A task which monitors for updates from the `StorageManager` and takes
+    /// actions based on those updates, such as informing Nexus and setting
+    /// up dump locations.
+    pub storage_monitor: StorageMonitorHandle,
+
+    /// A mechanism for interacting with the hardware device tree
+    pub hardware_manager: HardwareManager,
+
+    /// A mechanism for interacting with the task that monitors for hardware
+    /// updates from the [`HardwareManager`]
+    pub hardware_monitor: HardwareMonitorHandle,
+
+    // A handle for interacting with the bootstore
+    pub bootstore: bootstore::NodeHandle,
+
+    // A reference to the object used to manage zone bundles
+    pub zone_bundler: ZoneBundler,
+}
+
+/// Spawn all long running tasks
+pub async fn spawn_all_longrunning_tasks(
+    log: &Logger,
+    sled_mode: SledMode,
+    global_zone_bootstrap_ip: Ipv6Addr,
+    config: &Config,
+) -> LongRunningTaskHandles {
+    let storage_key_requester = spawn_key_manager(log);
+    let mut storage_manager =
+        spawn_storage_manager(log, storage_key_requester.clone());
+
+    let storage_monitor = spawn_storage_monitor(log, storage_manager.clone());
+
+    // TODO: Does this need to run inside tokio::task::spawn_blocking?
```

Comment:
I would vote "yes", just from a "don't do synchronous potentially-blocking stuff on tokio worker threads" point of view. At a minimum this makes a handful of system calls (both to `sysconf` and to poll the device tree).

Reply:
Done in a7848f91f